libnavigation-core/src/test/java/com/mapbox/navigation/core/routerefresh/RouteRefreshControllerTest.kt

```diff
@@ -127,37 +153,82 @@ class RouteRefreshControllerTest {
+        runBlockingTest {
+            val primaryRoute = createNavigationRoute(createTestTwoLegRoute(requestUuid = null))
+            val routeRefreshController = createRouteRefreshController()
 
-        val refreshedDeferred = async { routeRefreshController.refresh(listOf(primaryRoute)) }
-        advanceTimeBy(TimeUnit.HOURS.toMillis(6))
+            val refreshedDeferred = async { routeRefreshController.refresh(listOf(primaryRoute)) }
+            advanceTimeBy(TimeUnit.HOURS.toMillis(6))
 
-        assertTrue(refreshedDeferred.isActive)
-        verify(exactly = 1) {
-            logger.logI(
-                withArg {
-                    assertTrue(
-                        "message doesn't mention the reason of failure - empty uuid: $it",
-                        it.contains("uuid", ignoreCase = true)
+            assertTrue(refreshedDeferred.isActive)
+            verify(exactly = 1) {
+                logger.logI(
+                    withArg {
+                        assertTrue(
+                            "message doesn't mention the reason of failure - empty uuid: $it",
+                            it.contains("uuid", ignoreCase = true)
+                        )
+                    },
+                    any()
+                )
+            }
+            refreshedDeferred.cancel()
+            verify(ordering = Ordering.ORDERED) {
+                mockStatesObserver.onNewState(
+                    RouteRefreshStateResult(
+                        RouteRefreshExtra.REFRESH_STATE_FINISHED_FAILED,
+                        "No routes which could be refreshed. null#0 DirectionsRoute#requestUuid " +
+                            "is blank. This can be caused by a route being generated by an " +
+                            "Onboard router (in offline mode). Make sure to switch to an " +
+                            "Offboard route when possible, only Offboard routes support " +
+                            "the refresh feature."
                     )
-                },
-                any()
-            )
+                )
+            }
         }
-        refreshedDeferred.cancel()
-    }
 
     @Test
-    fun `refreshing of empty routes`() = runBlockingTest {
+    fun `refreshing of empty routes, observer is not triggered`() = runBlockingTest {
         val routeRefreshController = createRouteRefreshController()
 
-        val refreshedDeferred = async { routeRefreshController.refresh(listOf()) }
+        val refreshedDeferred = async {
+            routeRefreshController.refresh(
+                listOf()
+            )
+        }
         advanceTimeBy(TimeUnit.HOURS.toMillis(6))
 
         assertTrue(refreshedDeferred.isActive)
         refreshedDeferred.cancel()
+
+        verify(exactly = 0) {
+            mockStatesObserver.onNewState(any())
+        }
+    }
+
+    @Test
+    fun `refresh canceled, observer is not triggered`() = runBlockingTest {
```

Comment:
Shouldn't we have some closing state delivered for each start state? Should we introduce a `CANCELED` state next to existing `SUCCESS` and `FAILURE`?

Reply:
At this particular case refresh canceled at the moment when actual refresh hasn't been stared, so `observer is not triggered` is expected behavior. 
But in common it makes sense. I'll add one more state